### PR TITLE
Feature/5714 dynamic orgs

### PR DESCRIPTION
**Pull request by v-dumas**
Allow to specify the target organizations of a Communication through an OQL

### CODE REVIEW

File: datamodel.itop-communications.xml [hunk 1]
```diff
@@ -232,33 +244,40 @@
...
-				$bRet = in_array($iUserOrg, $aTargetOrgs);
-			}
-			else // 'cascade'
-			{
-				// Is the user org below or within any of the target organizations
-				$oOrgHierarchical = DBSearch::FromOQL_AllData('SELECT Organization AS child JOIN Organization AS root ON child.parent_id BELOW root.id WHERE root.id IN (:target_org_list) AND child.id = :contact_org_id');
-				$oOrgHSet = new DBObjectSet($oOrgHierarchical, array(), array('contact_org_id' => $iUserOrg, 'target_org_list' => $aTargetOrgs));
-				$bRet = $oOrgHSet->Count() > 0;
-			}
+		if ($sMatchType == 'oql') {
```
Comment (Molkobain) [inline datamodel.itop-communications.xml:254]:
```suggestion
		if ($sMatchType === 'oql') {
```

We tend to prefer strict comparison when possible (= variable type is known to alsays be the same), this avoid regression in the future.

File: datamodel.itop-communications.xml [hunk 2]
```diff
@@ -232,33 +244,40 @@
...
+		else {
+		    // Target organizations are those linked to the Communication
+			$oOrgSearch = DBSearch::FromOQL_AllData('SELECT lnkCommunicationToOrganization WHERE communication_id = :communication_id');
+		    $oOrgSet = new DBObjectSet($oOrgSearch, array(), array('communication_id' => $this->GetKey()));
+		    $aTargetOrgs = $oOrgSet->GetColumnAsArray('org_id');
+		    if (count($aTargetOrgs) == 0) {
+		        // Target organizations are "All organizations" when none are explicitly linked to the Communication
+			    $bRet = true;
+		    }
+		    elseif ($sMatchType == 'direct') {
```
Comment (Molkobain) [inline datamodel.itop-communications.xml:270]:
```suggestion
		    elseif ($sMatchType === 'direct') {
```

File: datamodel.itop-communications.xml [hunk 3]
```diff
@@ -232,33 +244,40 @@
...
+    			// Target organizations are those directly linked to the communication
+    			$bRet = in_array($iUserOrg, $aTargetOrgs);
+			}
+    		else { // ($sMatchType == 'cascade')
```
Comment (Molkobain) [inline datamodel.itop-communications.xml:274]:
```suggestion
    		else { // ($sMatchType === 'cascade')
```

File: datamodel.itop-communications.xml [hunk 4]
```diff
@@ -299,13 +318,73 @@
...
-    parent::DoCheckToWrite();
-    if ($this->Get('start_date') > $this->Get('end_date'))
     {
-      $this->m_aCheckIssues[] = Dict::Format('Class:Communication/Error:EndDateMustBeGreaterThanStartDate');
+      parent::DoCheckToWrite();
+      if ($this->Get('start_date') > $this->Get('end_date')) {
+          $this->m_aCheckIssues[] = Dict::S('Class:Communication/Error:EndDateMustBeGreaterThanStartDate');
+      }
+      $sOql = $this->Get('org_oql');
+      if ($this->Get('org_match_type') == 'oql') {
```
Comment (Molkobain) [inline datamodel.itop-communications.xml:327]:
```suggestion
      if ($this->Get('org_match_type') === 'oql') {
```

File: datamodel.itop-communications.xml [hunk 5]
```diff
@@ -299,13 +318,73 @@
...
+          if (empty($sOql)) {
+              $this->m_aCheckIssues[] = Dict::S('Class:Communication/Error:OQLMustBeProvided');
+          } else {
+              try {
+                  $oSearch = DBObjectSearch::FromOQL($sOql);
+                  $sOqlClass = $oSearch->GetClass();
+                  if ($sOqlClass != 'Organization') {
```
Comment (Molkobain) [inline datamodel.itop-communications.xml:334]:
```suggestion
                  if ($sOqlClass !== 'Organization') {
```

**Review by Molkobain — CHANGES_REQUESTED** [review]

File: datamodel.itop-communications.xml (hunk 3 above)
Comment (v-dumas) [reply datamodel.itop-communications.xml:274]: Fix a test in a comment, you're aiming for perfection ;-)

**Review by Molkobain — APPROVED** [review]

**Pull request merged**